src/provider/embedded.rs

```diff
@@ -0,0 +1,462 @@
+//! An embedded database backed `Provider` implementation.
+//!
+//! Currently, the underlying storage engine uses the [sled embedded
+//! database](https://github.com/spacejam/sled). This database provides caching support as well as
+//! atomic operations. Underneath the hood, we encode the data using the
+//! [CBOR](https://github.com/pyfisch/cbor) format for efficient serialization/deserialization from
+//! the database.
+//!
+//! This provider is currently experimental, with the goal of replacing the `FileProvider` as the
+//! default provider in the future. This provider is only available with the `embedded` feature
+//! enabled. Please note that the embedded data store and encoding format may change until this
+//! feature is stabilized
+
+use std::convert::TryInto;
+use std::path::Path;
+use std::sync::Arc;
+
+use sha2::{Digest, Sha256};
+use sled::Error as SledError;
+use tokio::io::AsyncReadExt;
+use tokio::sync::Semaphore;
+use tokio_stream::{Stream, StreamExt};
+use tokio_util::codec::{BytesCodec, FramedRead};
+use tokio_util::io::StreamReader;
+use tracing::{debug, error, info, instrument, trace, warn};
+use tracing_futures::Instrument;
+
+use crate::provider::{Provider, ProviderError, Result};
+use crate::search::Search;
+use crate::verification::Verified;
+use crate::{Id, Signed};
+
+const INVOICE_DB_NAME: &str = "invoices";
+const PARCEL_DB_NAME: &str = "parcels";
+// TODO: This number should be equal to the number of threads configured for blocking. We could
+// expose this value in the constructor, but that feels too much like a low-level detail to expose
+// in the API. But I also can't find a way to fetch this configured value
+const BLOCKING_THREAD_COUNT: usize = 512;
+
+/// An embedded database backend for storing and retrieving bindles and parcles.
+///
+/// Given a storage directory, EmbeddedProvider brings its own storage layout for keeping track of
+/// Bindles.
+///
+/// An EmbeddedProvider needs a search engine implementation. When invoices are created or yanked,
+/// the index will be updated.
+pub struct EmbeddedProvider<T> {
+    invoices: sled::Tree,
+    parcels: sled::Tree,
+    index: T,
```

Comment:
This is interesting... we could actually implement search indices in Sled as well, which would mean we could persist search indices. For full (non-strict) searches, we could implement something really robust! This would be great!